[MLIR][DISC] legalize tensor_load inserted during hlo-to-lhlo conversion

wyzero opened the pull request:
This PR implements logic for lowering memref.tensor_load ops that are
inserted during `mhlo-legalize-to-lmhlo`

sanjoy commented on tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:
```diff
@@ -1651,7 +1660,9 @@ def HLO_ReshapeOp: HLO_Op<"reshape",
 
-def HLO_DynamicReshapeOp: HLO_Op<"dynamic_reshape", [NoSideEffect]> {
+def HLO_DynamicReshapeOp: HLO_Op<"dynamic_reshape", [NoSideEffect,
+    DeclareOpInterfaceMethods<InferShapedTypeOpInterface,
+        ["inferReturnTypeComponents", "reifyReturnTypeShapes"]>]> {
```
Can we avoid repeating this boilerplate somehow?

wyzero replied:
One possible solution is that putting these interfaces into the base op and provide a default implementation (e.g. return failure directly). Is there any suggestion about this?

sanjoy commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:
```diff
@@ -2348,6 +2696,71 @@ static LogicalResult Verify(DynamicPadOp op) {
+  return failure();
+}
+
+LogicalResult DynamicPadOp::reifyReturnTypeShapes(
```
Where is this function (`reifyReturnTypeShapes`) tested?

wyzero replied:
This filecheck-based ut will test this:
https://github.com/tensorflow/tensorflow/pull/49598/files#diff-a6fc1c1e273cf5234d691f9a691f08b8cfbc9c07107558dda12a2009606d7fd6R56

sanjoy reviewed (changes requested):
I'd suggest breaking up this PR into smaller PRs and introduce the dynamic ops one or two at a time.  That'll help ensure a more thorough code review.

wyzero commented:
> I'd suggest breaking up this PR into smaller PRs and introduce the dynamic ops one or two at a time. That'll help ensure a more thorough code review.

Thanks. I'll take a try.

Commit:
[MLIR][DISC] legalize tensor_load inserted during hlo-to-lhlo conversion

This PR implements logic for lowering memref.tensor_load ops that are
inserted during `mhlo-legalize-to-lmhlo`.

wyzero commented:
@sanjoy Hi, I split the original PR into small PRs and sent the first PR. Could you take a look again?

joker-eph commented on tensorflow/compiler/mlir/hlo/tests/lhlo-legalize-tensor-load-op.mlir:
```diff
@@ -0,0 +1,34 @@
+  %c1 = constant 1 : index
+  %c2 = constant 2 : index
+  // CHECK-NOT: memref.tensor_load
+  // CHECK-NOT: tesnor.extract
```
```suggestion
  // CHECK-NOT: tensor.extract
```

wyzero replied:
Thanks! Done.

joker-eph commented:
LGTM otherwise.

Commit:
minor fix

wyzero commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc:
```diff
@@ -0,0 +1,99 @@
+                                           tensor_load.memref());
+    return success();
+  }
+};
```
Thanks! These patterns are relative specific to hlo-to-lhlo conversion pass. I am not sure if there are other places needing these patterns. I think we can leave it here for now and upstream them if necessary in the future (I'd like to help).

joker-eph replied:
Can you elaborate what's specific to HLO here?

wyzero replied:
Interface `InferReturnTypeShapes` of mhlo ops are implemented on tensor level. During mhlo-to-lmhlo conversion,  the operands of a hlo op are legalized to memref type first before the op is legalized,  making it invalid to call `InferReturnTypeShapes` directly.  To solve this problem, we need to insert tensor_load op to make sure the operands passed in to the interface are valid.

This pr tries to remove these tensor_load ops after conversion in some cases. I'm not sure if these patterns are likely to be used in other places since it's highly related to the implementation of tensor-to-buffer conversion. However I agree that the implementation of these patterns has nothing to do with hlo.

sanjoy commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc:
```diff
@@ -0,0 +1,99 @@
+  auto context = &getContext();
+  OwningRewritePatternList patterns(context);
+  patterns.insert<ForwardShapeOfOp, ForwardExtractOp>(context);
+  (void)applyPatternsAndFoldGreedily(getFunction(), std::move(patterns));
```
Is it OK to discard the return value here?  If it is always supposed to return `success` should we `assert` that here?

wyzero replied:
Done!

sanjoy replied:
I don't see any changes here?

wyzero replied:
Oh, sorry, I did miss it here.

sanjoy commented on tensorflow/compiler/mlir/hlo/tests/lhlo-legalize-tensor-load-op.mlir:
```diff
@@ -0,0 +1,34 @@
+  %c2 = constant 2 : index
+  // CHECK-NOT: memref.tensor_load
+  // CHECK-NOT: tensor.extract
+  // CHECK: memref.load
+  // CHECK: memref.load
+  // CHECK: memref.load
```
You should also check that the indices are correct.

wyzero replied:
Done!

sanjoy reviewed (changes requested):
Only very minor comments.

Commit:
minor refine

wyzero commented:
> Only very minor comments.

I have refined the code according to the comments. Could you have a look again?

wyzero commented:
@sanjoy Could you refresh the page? I have changed both in the last commit.

https://github.com/tensorflow/tensorflow/pull/49598/files#diff-2aece6a81b6b40b767c54d4762d4da714b85bd2ee4f99ec52cf96349ea0f2cc7R240

sanjoy commented:
> @sanjoy Could you refresh the page? I have changed both in the last commit.
> 
> https://github.com/tensorflow/tensorflow/pull/49598/files#diff-2aece6a81b6b40b767c54d4762d4da714b85bd2ee4f99ec52cf96349ea0f2cc7R240

I believe in that commit you changed tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc, LMK if I'm misunderstanding.

wyzero commented:
> > @sanjoy Could you refresh the page? I have changed both in the last commit.
> > [#49598 (files)](https://github.com/tensorflow/tensorflow/pull/49598/files#diff-2aece6a81b6b40b767c54d4762d4da714b85bd2ee4f99ec52cf96349ea0f2cc7R240)
> 
> I believe in that commit you changed tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc, LMK if I'm misunderstanding.

Sorry for that! you are right. I changed the wrong place. I pushed a new commit now.

wyzero commented:
@sanjoy Could you take a look again?

sanjoy commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc:
```diff
@@ -0,0 +1,101 @@
+  OwningRewritePatternList patterns(context);
+  patterns.insert<ForwardShapeOfOp, ForwardExtractOp>(context);
+  if (failed(applyPatternsAndFoldGreedily(func, std::move(patterns))))
+    return signalPassFailure();
```
Maybe drop the `return ` since `signalPassFailure` returns `void`?

wyzero replied:
Done!

sanjoy reviewed (changes requested):
One very minor nit (thanks for your patience!).

Commit:
minor fix

wyzero commented:
@sanjoy Thanks! I have changed it.

gbaned commented:
@wyzero  Can you please address Ubuntu Sanity errors? Thanks!

Commit:
fix

wyzero commented:
@gbaned Hi, I fix the error. Could you help to test again?

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc:
```diff
@@ -0,0 +1,101 @@
+  OwningRewritePatternList patterns(context);
+  patterns.insert<ForwardShapeOfOp, ForwardExtractOp>(context);
+  if (failed(applyPatternsAndFoldGreedily(func, std::move(patterns))))
+    signalPassFailure();
```
This may fail the pipeline silently, can you emit an error? (I think the only source of failure here is a non-convergence of the greedy driver)

wyzero replied:
Done!

Commit:
fix

wyzero commented:
@joker-eph Hi, I emit an error when failure. Could you help to take a look again?

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc:
```diff
@@ -0,0 +1,103 @@
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<shape::ShapeDialect, tensor::TensorDialect,
+                    memref::MemRefDialect>();
+  }
```
This does not seem needed. You only need to list here the dialects that are introduced by the pass and *not matched in the input*

wyzero replied:
Done

wyzero commented:
@gbaned Hi, It seems that the error in `Ubuntu Sanity errors` is not related to the part of code I changed. Is there anything I need to do (e.g. rebase to latest master)?   Or simply re-run the test again is enough?

```
FAIL: Found 10 errors
tensorflow/python/lib/io/file_io_test.py:673: [E1123(unexpected-keyword-arg), FileIoTest.testFileSeekableWithZip] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/mnist.py:76: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/reuters.py:114: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/boston_housing.py:62: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/imdb.py:108: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/tools/saved_model_cli.py:663: [E1123(unexpected-keyword-arg), load_inputs_from_input_arg_string] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/training/tracking/python_state_test.py:131: [E1123(unexpected-keyword-arg), _NumpyWrapper.deserialize] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/kernel_tests/random/multinomial_op_big_test.py:44: [E1123(unexpected-keyword-arg), MultinomialTest.testLargeDynamicRange] Unexpected keyword argument 'return_counts' in function call
tensorflow/python/kernel_tests/random/multinomial_op_big_test.py:62: [E1123(unexpected-keyword-arg), MultinomialTest.testLargeDynamicRange2] Unexpected keyword argument 'return_counts' in function call
tensorflow/python/kernel_tests/random/multinomial_op_big_test.py:85: [E1123(unexpected-keyword-arg), MultinomialTest.testLargeDynamicRange3] Unexpected keyword argument 'return_counts' in function call
```

Commit:
fix

joker-eph commented:
> @gbaned Hi, It seems that the error in Ubuntu Sanity errors is not related to the part of code I changed. Is there anything I need to do (e.g. rebase to latest master)? Or simply re-run the test again is enough?

It happens, either way is fine. Ignoring is fine as well: we re-run all of this when we try to integrate internally and we are used to re-trigger these builds when they fail for unrelated reasons.

gbaned commented:
@wyzero  Can you please address Ubuntu Sanity errors? Thanks!

wyzero commented:
> @wyzero Can you please address Ubuntu Sanity errors? Thanks!

As I said above, It seems that the error is not related to the part of code I changed. I will try to rebase to the latest master to see if this can work.

```
FAIL: Found 10 errors
tensorflow/python/lib/io/file_io_test.py:673: [E1123(unexpected-keyword-arg), FileIoTest.testFileSeekableWithZip] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/mnist.py:76: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/reuters.py:114: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/boston_housing.py:62: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/keras/datasets/imdb.py:108: [E1123(unexpected-keyword-arg), load_data] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/tools/saved_model_cli.py:663: [E1123(unexpected-keyword-arg), load_inputs_from_input_arg_string] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/training/tracking/python_state_test.py:131: [E1123(unexpected-keyword-arg), _NumpyWrapper.deserialize] Unexpected keyword argument 'allow_pickle' in function call
tensorflow/python/kernel_tests/random/multinomial_op_big_test.py:44: [E1123(unexpected-keyword-arg), MultinomialTest.testLargeDynamicRange] Unexpected keyword argument 'return_counts' in function call
tensorflow/python/kernel_tests/random/multinomial_op_big_test.py:62: [E1123(unexpected-keyword-arg), MultinomialTest.testLargeDynamicRange2] Unexpected keyword argument 'return_counts' in function call
tensorflow/python/kernel_tests/random/multinomial_op_big_test.py:85: [E1123(unexpected-keyword-arg), MultinomialTest.testLargeDynamicRange3] Unexpected keyword argument 'return_counts' in function call
```

wyzero commented:
@gbaned Could you test again?

wyzero commented:
@gbaned  It's still the same error. It seems that pylint check failed. However I don't change any python files at all. Is anyone else meet the same problem? 

```
==== Summary of sanity check results ====
1. do_configure_test: Run ./configure
  PASS
2. do_pylint: Python 3 pylint
  FAIL
3. do_buildifier: buildifier check
  PASS
4. do_bazel_nobuild: bazel nobuild
  PASS
5. do_bazel_deps_query: bazel query
  PASS
6. do_pip_package_licenses_check: pip: license check for external dependencies
  PASS
7. do_lib_package_licenses_check: C library: license check for external dependencies
  PASS
8. do_java_package_licenses_check: Java Native Library: license check for external dependencies
  PASS
9. do_pip_smoke_test: Pip Smoke Test: Checking py_test dependencies exist in pip package
  PASS
10. do_check_load_py_test: Check load py_test: Check that BUILD files with py_test target properly load py_test
  PASS
11. do_code_link_check: Code Link Check: Check there are no broken links
  PASS
12. do_check_file_name_test: Check file names for cases
  PASS
13. do_pip_no_cuda_deps_check_ubuntu: Check Ubuntu gpu pip package does not depend on cuda shared libraries
  PASS
14. do_pip_no_cuda_deps_check_windows: Check Windows gpu pip package does not depend on cuda shared libraries
  PASS
```

wyzero commented:
@gbaned @joker-eph @sanjoy Could trigger the test again?

wyzero commented:
@joker-eph @sanjoy Hi, Could approve this PR again?

joker-eph commented:
> @joker-eph @sanjoy Hi, Could you approve this PR again?

It does not need a re-approval, it is in the integration pipeline since the original approvals, it didn't go through by itself because the BUILD file is missing some dependencies that I had to fix manually (I don't know why Bazel is OK with this but it failed internally)

wyzero commented:
@joker-eph Thanks! Is there anything I can do to accelerate this process? It would be nice if we can finish this PR quickly since we still have a lot of new PRs in the queue.

joker-eph commented:
> @joker-eph Thanks! Is there anything I can do to accelerate this process? It would be nice if we can finish this PR quickly since we still have a lot of new PRs in the queue.

No: I had to rebase multiple times to work around issues with the various bots...
(also Google was closed last Friday, and Monday was a holiday in the US, many people to a day or day two extra to make it almost a week off!)
